common/comms/src/inbound_message_service/message_dispatcher.rs

```diff
@@ -0,0 +1,146 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use std::collections::HashMap;
+
+#[derive(Debug, Error)]
+pub enum DispatchError {
+    /// A dispatch route was not defined for the specific message type
+    MessageHandlerUndefined,
+}
+
+/// Each message struct that will be dispatched should implement Dispatchable. The dispatch_type function is used by the
+/// MessageDispatcher to be able to distinguish between different message types and select which handler function should
+/// process the message.
+pub trait Dispatchable {
+    fn dispatch_type(&self) -> u32;
```

Comment:
u32? Why not a custom enum? (These can resolve to u32's)
https://doc.rust-lang.org/rust-by-example/custom_types/enum/c_like.html

Reply:
The ```MessageDispatcher``` is a generic dispatcher that can be used to setup comms and/or domain level dispatching allowing the dispatching to be configured differently depending on the type of node or client. The u32 is used because the message types Enum will be different for each dispatcher and we don't have to fix it to a specific set of types allowing the implementation to select its own types depending on which dispatcher is being implemented. As an example the tests and dispatcher provided in ```comms_msg_handler.rs``` each has their own dispatch types enum that is used to dispatch.

Reply:
I feel like this approach could lead to tough to track bugs that Rust's type system could be leveraged to avoid; but ok.